machine/api.go

```diff
@@ -0,0 +1,217 @@
+// Package machine keeps track of machines and applications for scotty.
+package machine
+
+import (
+	"github.com/Symantec/Dominator/lib/mdb"
+	"github.com/Symantec/scotty"
+	"github.com/Symantec/scotty/application"
+	"github.com/Symantec/scotty/awsinfo"
+	"github.com/Symantec/scotty/namesandports"
+	"github.com/Symantec/scotty/store"
+	"sort"
+	"sync"
+	"time"
+)
+
+// EndpointObservation represents an observation of applications and ports
+// running on a certain machine.
+type EndpointObservation struct {
+	// The observation number that increases by 1 each time
+	SeqNo uint64
+
+	// The name of each appliation and the port it is running on.
+	Endpoints namesandports.NamesAndPorts
+}
+
+// EndpointObservations stores endpoint observations for all machines.
+// EndpointObservations is safe to use with multiple goroutines.
+type EndpointObservations struct {
+	mu   sync.Mutex
+	data map[string]EndpointObservation
+}
+
+// NewEndpointObservations returns an empty EndpointObservations
+func NewEndpointObservations() *EndpointObservations {
+	return &EndpointObservations{data: make(map[string]EndpointObservation)}
+}
+
+// Save saves namesAndPorts for hostName and increments the sequence number
+// by 1.
+func (e *EndpointObservations) Save(
+	hostName string, namesAndPorts namesandports.NamesAndPorts) {
+	e.save(hostName, namesAndPorts)
+}
+
+// GetAll returns a snapshot of all the endpoint observations.
+func (e *EndpointObservations) GetAll() map[string]EndpointObservation {
+	return e.getAll()
+}
+
+// MaybeAddApp atomically adds an application for hostName if its port
+// isn't there already. If nothing has been previously stored for hostName,
+// MaybeAddApp does nothing.
+func (e *EndpointObservations) MaybeAddApp(hostName, appName string, port uint) {
+	e.maybeAddApp(hostName, appName, port)
+}
+
+// Machine represents a single machine
+type Machine struct {
+
+	// The host name of the machine
+	Host string
+
+	// Whether or not machine is active
+	Active bool
+
+	// Aws information for machine. nil if no aws information available.
+	Aws *awsinfo.AwsInfo
+}
+
+// AccountId is a convenience routine that returns the aws account id.
+// If no aws information, returns the empty string.
+func (m *Machine) AccountId() string {
```

Comment:
Similar to my other comments, we shouldn't have AWS specific constructs in a generic package. We should have generic constructs.